proposals/0339-replace-pthread-implementation-with-threads-from-cpp11.md

```diff
@@ -65,24 +65,88 @@ private:
 
   std::thread thread_;
 };
-  ```
-
-The same changes should be introduced not only for the threads but also for the synchronization primitives like mutex, conditional variable, etc., where required.
+```
 
 
 ## Potential downsides
 
 C++11 provides no equivalent to `pthread_cancel(pthread_t thread)`.
 As an option to resolve this issue, we can add some `cancellation_token` to thread attributes which contains a sign that the thread has been cancelled.
 
-Pthreads provides control over the size of the stack of created threads; C++11 does not address this issue.
-`std::thread::native_handle` can be used if needed.
 
+Here is an simplified example of how SDL Сore will implement stopping threads:
+```
+/*   Cancellation Token Class  */
+
+class CancellationToken {
+public:
+  CancellationToken() : _cancelled(false) {}
+  operator bool() const { return !_cancelled; }
+  void cancel() { _cancelled = true; }
+private:
+  std::atomic<bool> _cancelled;
+};
+
+/*  Thread Wrapper Class */
+
+class ThreadWrapper {
+protected:
+  std::thread _thread;
+  CancellationToken _token;
+
+  void Reset() {
+    if (!_thread.joinable())
+      return;
+
+    _token.cancel();
+    _thread.join();
+  }
+
+public:
+  ThreadWrapper() = default;
+
+  template <typename Delegate, typename... Args>
+  void Start(Delegate &&delegate, Args... args) {
+    _thread = std::thread(delegate, args..., std::ref(_token));
+  }
+
+  void Stop() {
+    Reset();
+  }
+
+  ~ThreadWrapper() { Reset(); }
+};
+
+/*    Test Class Using Thread Wrapper  */
+
+class UsbHandler : public ThreadWrapper {
+public:
+  void Init() {
+    Start([](CancellationToken &token) {
+      while (token) {
+        ....
+      }
+    });
+  }
+};
+
+/*    Usage   */
+void Test() {
+  UsbHandler uh;
+  uh.Init();
+
+  std::this_thread::sleep_for(std::chrono::seconds(5));
+  uh.Stop();
+}
+```
+
+Pthreads provides control over the size of the stack of created threads; C++11 standard threads does not address this issue.
+However in SDL Core we don't need to specify a stack size. Since the default stack size meets SDL Core requirements.
 
 
 ## Impact on existing code
 The thread approach of SDL Core will be impacted.
-Since this is just a refactoring of existing code without adding new features, there is no need to change a minor version of the library.
+Since this will affect the classes of multithreaded code, a major version change of SDL Core is required.
```

Comment:
per http://www.semver.org/ 
> MAJOR version when you make incompatible API changes

I do not see the proposed API changes in the proposal
Could you please elaborate on this?

Reply:
Should we say 
```
The thread approach of SDL Core will be impacted. Removing the current thread implementation, delegate classes, and public methods will be done, a major version change of SDL Core is required.
```

Reply:
@KhrystynaDubovyk 
Done in f1ef850da4688559b8e54b16563d1c3bccbcfb4d